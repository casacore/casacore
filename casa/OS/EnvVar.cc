//# EnvVar.cc: Environment variables class
//# Copyright (C) 1993,1994,1995,1998,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/OS/EnvVar.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>
#include <casa/stdlib.h>
#include <cstring>                  //# for strcpy with gcc-4.3


namespace casa { //# NAMESPACE CASA - BEGIN

Bool EnvironmentVariable::isDefined (const String& name)
{
  return getenv (name.chars());
}

String EnvironmentVariable::get (const String& name)
{
  Char* env = getenv (name.chars());
  if (env) return String(env);
  return String();
}

void EnvironmentVariable::set (const String& name, const String& value)
{
  uInt nl = name.length();
  uInt vl = value.length();
  Char* str = new Char [nl + vl + 2];
  strcpy (str, name.chars());
  str[nl] = '=';
  strcpy (str+nl+1, value.chars());
  // Note that putenv takes over the pointer, so we should not delete str.
  AlwaysAssert (putenv(str) == 0, AipsError);
}

} //# NAMESPACE CASA - END

