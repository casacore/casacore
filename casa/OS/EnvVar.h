//# EnvVar.h: Environment variables class
//# Copyright (C) 1993,1994,1995,1999,2002
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_ENVVAR_H
#define CASA_ENVVAR_H

#include <casacore/casa/aips.h>
#include <casacore/casa/BasicSL/String.h>

namespace casa { //# NAMESPACE CASA - BEGIN

// <summary>
// This class allows for getting enviroment variables
// </summary>

// <use visibility=export>
// <reviewed reviewer="Paul Shannon" date="1995/02/08" tests="tEnvVar" demos="">

// <synopsis> 
// Environment variables are familiar to every Unix, MSDOS and VMS
// computer user.  This class makes it convenient to get and
// enquire about environment variables from within a C++ program.
// </synopsis> 

// <example>
// Check if an environment variable is defined.
// If so, get its value.
// <srcblock>
//  if (EnvironmentVariable::isDefined ("PATH")) {
//    cout << EnvironmentVariable::get ("PATH") << endl;
//  }
// </srcblock>
// </example>

// <todo asof=1994/02/08>
// </todo>

class EnvironmentVariable
{
public:
  // Is environment variable with given name defined? 
  static Bool isDefined (const String& name);

  // Get the value of environment variable with given name.
  // If not defined, return an empty String.
  static String get (const String& name);

  // Define environment variable.
  // If it already exists, its value will be overwritten.
  static void set (const String& name, const String& value);

private:
  // This class is not meant to be constructed.
  EnvironmentVariable();
};



} //# NAMESPACE CASA - END

#endif

