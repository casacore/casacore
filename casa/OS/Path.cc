//# Path.cc: Class to define a pathname
//# Copyright (C) 1993,1994,1995,1996,1997,1998,1999,2000,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <casa/OS/Path.h>
#include <casa/OS/EnvVar.h>
#include <casa/Arrays/Vector.h>
#include <casa/Arrays/ArrayUtil.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions.h>

#include <pwd.h>                    // needed for getpwnam
#include <unistd.h>                 // needed for pathconf
#include <limits.h>                 // needed for PATH_MAX, etc.
#include <ctype.h>                  // needed for isprint


namespace casa { //# NAMESPACE CASA - BEGIN

// The maximum number of bytes in a pathname is 255 (_POSIX_PATH_MAX)
// Definition for POSIX systems
#if defined (_POSIX_PATH_MAX)
const uInt pathmax_posix = _POSIX_PATH_MAX;
#else
const uInt pathmax_posix = 255;
#endif


// The maximum number of bytes in a pathname is PATH_MAX 
#if defined (PATH_MAX)
static uInt pathMax = PATH_MAX;
#else
static uInt pathMax = 0;
#endif

const uInt PATH_MAX_GUESS = 1024; // if PATH_MAX is indeterminate
                                  // we're not guaranteed this is adequate. 

// The maximum number of bytes in a filename is 14 (_POSIX_NAME_MAX)
// Definition for POSIX systems
#if defined (_POSIX_NAME_MAX)
const uInt namemax_posix = _POSIX_NAME_MAX;
#else
const uInt namemax_posix = 14;
#endif


// The maximum number of bytes in a filename is NAME_MAX
#if defined (NAME_MAX)
static uInt nameMax = NAME_MAX;
#else
static uInt nameMax = 0;
#endif

const uInt NAME_MAX_GUESS = 255;  // if NAME_MAX is indeterminate
                                  // we're not guaranteed this is adequate. '


Path::Path() 
: itsOriginalPathName (".")
{}

Path::Path (const String& pathName)
: itsOriginalPathName (pathName)
{
    if (itsOriginalPathName.empty()) { 
	itsOriginalPathName = ".";    
    }
}    

Path::Path (const Path& that)
: itsOriginalPathName (that.itsOriginalPathName)
{}

Path::~Path()
{}   

Path& Path::operator= (const Path& that)
{
    if (this != &that) {   
	itsOriginalPathName = that.itsOriginalPathName;
	itsAbsolutePathName = that.itsAbsolutePathName;
	itsExpandedPathName = that.itsExpandedPathName;
    }
    return *this;
}

void Path::append (const String& string)
{
    if  (!string.empty()) {
	if (itsOriginalPathName.lastchar() != '/'
        &&  string.firstchar() != '/') {
	    itsOriginalPathName += "/";
	}
	itsOriginalPathName += string;
	itsAbsolutePathName = "";
	itsExpandedPathName = "";
    }
}

const String& Path::expandedName() const
{
    if (itsExpandedPathName.empty()) {
	itsExpandedPathName = expandName (itsOriginalPathName);
    }
    return itsExpandedPathName;
}

const String& Path::absoluteName() const
{
    if (itsAbsolutePathName.empty()) {
	itsAbsolutePathName = removeDots (makeAbsoluteName (expandedName()));
    }
    return itsAbsolutePathName;
}

Bool Path::isValid() const
{
    // Check if the length of the pathname is not too long
    if (itsOriginalPathName.length() > getMaxPathNameSize()) {
	return False;
    }
    // Check if pathname contains double slashes
    if (itsOriginalPathName.contains ("//")) {
	return False;
    }
    // Check if pathname contains non-printables
    uInt i;
    for (i=0; i<itsOriginalPathName.length(); i++) {
	if (isprint (itsOriginalPathName[i]) == 0 ) {
	    return False;
	}
    }

    // Check if the length of the pathname is not too long
    if (itsOriginalPathName.length() > getMaxPathNameSize()) {
	return False;
    }

    // Check if filenames are not too long
    String subPathname[30];
    String sep = "/";
    uInt nw = split (itsOriginalPathName, subPathname, 15, sep);
    uInt nameSize = getMaxNameSize();
    for (i=0; i<nw; i++) {
	if (subPathname[i].length() > nameSize ) {
	    return False;
	}
    }
    return True;
}

Bool Path::isStrictlyPosix() const
{
    // Check if the length of the pathname is not too long, according POSIX 
    // standard
    if(itsOriginalPathName.length() > pathmax_posix ) {
	return False;
    }
    // Check if pathname contains double slashes
    if (itsOriginalPathName.contains ("//")) {
	return False;
    }
    // Check if pathname contains non-printables
    uInt i;
    for (i=0; i<itsOriginalPathName.length(); i++) {
	if (!isprint (itsOriginalPathName[i])) {
	    return False;
	}
    }

    // Check if filenames are not too long, according POSIX standard
    String subPathname[30];
    String sep = "/";
    uInt nw = split(itsOriginalPathName, subPathname, 15, sep);
    for (i=0; i<nw; ++i) {
	if (subPathname[i].length() > namemax_posix )
	    return False;
    }
    return True;
}

uInt Path::length() const
{
    return itsOriginalPathName.length();
}

uInt Path::maxLength() const
{
    return getMaxPathNameSize();
}

String Path::baseName() const
{
    // Determine from expanded path name.
    String name = expandedName();
    // Search last slash.
    // Get rid of trailing slash.
    Int len=name.length();
    if (len > 0  &&  name[len-1] == '/') {
	len--;
    }
    Int i=len;
    while (--i >= 0  &&  name[i] != '/') {}
    // The base name is the part from the slash till the end.
    return name(i+1, len-i-1);
}

String Path::dirName() const
{
    String name = expandedName();
    // Search last slash.
    // Get rid of trailing slash (except if name consists of a slash only).
    Int i=name.length();
    if (i > 1  &&  name[i-1] == '/') {
	i--;
    }
    while (--i >= 0  &&  name[i] != '/') {}
    // If no slash, the dirname is the current directory.
    if (i < 0) {
	return ".";
    }
    // If the slash found is not the first character, skip it.
    if (i > 0) {
	i--;
    }
    return name.through(i);
}


uInt Path::getMaxPathNameSize()
{
    // pathMax is not defined(<0) then pathconf sets pathMax, 
    // if this doesn't work pathMax will get the value of PATH_MAX_GUESS
    if (pathMax == 0) {
#if defined(AIPS_CRAY_PGI)
        pathMax = PATH_MAX_GUESS;
#else
	pathMax = pathconf ("/",_PC_PATH_MAX) < 0  ?  pathMax : PATH_MAX_GUESS;
#endif
    }
    return pathMax;
}

uInt Path::getMaxNameSize()
{
    // nameMax is not defined (<0) then pathconf sets nameMax, 
    // if this doesn't work nameMax will get the value of PATH_MAX_GUESS
    if (nameMax == 0) {
#if defined(AIPS_CRAY_PGI)
        pathMax = NAME_MAX_GUESS;
#else
	nameMax = pathconf ("/",_PC_NAME_MAX) < 0  ?  nameMax : NAME_MAX_GUESS;
#endif
    }
    return nameMax;
}


String Path::expandName (const String& inString) const
{
    String tempString (inString);
    uInt cursor = 0;
    uInt i = 0;
    Bool flag = True;
    uInt count = 0;
    // Flag is set True when an environment variable is detected. When this 
    // happens more then 25 times, there is probably a recursive variable set.
    // In that case an exception will be thrown.
    while (flag && count<25) {  
	// flag is False, when there is not an environment variable
	// the name will not be checked again
	flag = False;     
	count++;          // count is increased when the string is 
	cursor = 0;       // walked through 
	// Replace tilde with the name of the home directory
	if (tempString.firstchar() == '~') {
	    if (tempString.length() == 1  ||  tempString[1] == '/') {
	      // To get the homedirectory, environment variable HOME is used.
		String name (EnvironmentVariable::get("HOME"));
		if (! name.empty()) {
		    tempString.del ("~",0);
		    tempString.prepend (name);
		}
	    } else {
		tempString.del ("~",0);
		getNextName (tempString, cursor);
		String temp (tempString.before(Int(cursor)));
		// The password file is used to get the home directory 
		// of "~name"
		// This cannot be done on the CRAY XT3 CATAMOUNT as it
		// does not support sockets.
#ifdef AIPS_CRAY_PGI
		tempString.prepend ("~");
#else
		passwd* passWd = getpwnam(temp.chars());
		if (passWd != 0) {
		    tempString.del (tempString.before (Int(cursor)));
		    tempString.prepend (passWd->pw_dir);
		    cursor = 0;
		}else{
		    tempString.prepend ("~");
		}
#endif
	    }
	}
	cursor = 0;
	i = 0;
	while (i < tempString.length()) {
	    getNextName (tempString, i);      // i is set on the next name   
	    if (tempString[cursor] == '$') {  // Environment variable detected
		String dName (tempString.at (Int(cursor+1), Int((i-cursor)-1)));
		String name (EnvironmentVariable::get(dName));
		if (! name.empty()) {
		    String res (name);
		    res.prepend (tempString.before(Int(cursor)));
		    res += tempString.after (Int(i-1));
		    tempString = res;
		    // name is placed in the new string and the old name is 
		    // gone
		    // flag is set True, so the name will be checked again for 
		    // environment variables
		    flag = True;
		}
	    }	
	    cursor = i + 1;
	    i++;
	}
    }
    if (flag) {
	// An exception is thrown when the string is checked more then 25 times
	throw (AipsError ("Path::expandName: recursive environment variable"));
    }
    return tempString;
}


String Path::makeAbsoluteName (const String& inString) const
{
    // If the first char is a slash the name is already absolute.
    if (inString.firstchar() == '/') {
	return inString;
    }
    // Otherwise we have a relative pathname.
    // Remove a possible leading . or ./
    String workString (inString);
    if (workString.length() > 0) {
	if (workString[0] == '.') {
	    Int from(1);
	    if (workString.length() > 1){
               switch(workString[1]) {
                case '/' :
		   from = 2;
                   break;
                case '.' :
                   from = 0;
                   break;
               }
	    } 
	    workString = workString.from (from);
	}
    }
    // Get the working directory and prepend it.
    // Insert a / when needed.
    char temp[1024];
    getcwd (temp, 1024);
    String tempString (temp);
    // Return the working directory when no input string left.
    if (workString.empty()) {
	return tempString;
    }
    // Insert a / when needed.
    if (tempString.lastchar() != '/') {
	tempString += '/';
    }
    tempString += workString;
    return tempString;
}

String Path::removeDots (const String& inString) const
{
    // Split the name at the slashes.
    Vector<String> parts (stringToVector (inString, '/'));
    Vector<uInt> validParts (parts.nelements());
    String dot(".");
    String dotdot("..");
    uInt nvalid = 0;
    uInt i;
    // Count the number of valid parts and keep their index.
    // Ignore blanks and . parts.
    // A .. part removes an entry from the valid list (if possible).
    for (i=0; i<parts.nelements(); i++) {
        if (! (parts(i).empty()  ||  parts(i) == dot)) {
	    if (parts(i) == dotdot) {
	        if (nvalid > 0) {
		    nvalid--;
		}
	    } else {
	        validParts(nvalid++) = i;
	    }
	}
    }
    // Combine the parts into the output string.
    // Start it with a slash if the input started with a slash.
    String outString;
    Bool doSlash = (parts(0).empty());
    for (i=0; i<nvalid; i++) {
        if (doSlash) {
	    outString += '/';
	}
	outString += parts(validParts(i));
	doSlash = True;
    }
    return outString;
}

void Path::getNextName (const String& inString, uInt& count) const
{
    // Sets count on the next slash or on the end of the string
    Int inx = inString.index ("/", count);
    if (inx < 0) {
	count = inString.length();
    } else {
	count = inx;
    }
}


String Path::stripDirectory (const String& name, const String& otherName)
{
    // Add trailing slash if not there.
    String dir (Path(otherName).absoluteName());
    if (dir.lastchar() != '/') {
	dir += '/';
    }
    Int leng = dir.length();
    // Convert name to an absolute path name.
    String aName (Path(name).absoluteName());
    // If directory is contained in this name, return name without it.
    // Prepend by ././ indicating full name is removed.
    if (leng > 0) {
        Int aleng = aName.length();
	if (aleng > leng  &&  aName.before(leng) == dir) {
	    return "././" + aName.from (leng);
	} else {
	    // No match; see if name matches start of otherName.
	    // If so, append /. to remainder of otherName.
	    if (aleng+1 < leng  &&  dir.before(aleng+1) == aName+'/') {
	        return dir.from(aleng+1) + '.';
	    }
	    // No match; now compare using the directory part only.
	    dir = Path(dir).dirName() + '/';
	    while (dir.length() >= 2  &&  dir[0] == '.'  &&  dir[1] == '/') {
		dir = dir.after(1);
	    }
	    leng = dir.length();
	    if (leng > 0) {
		if (aName.length() > uInt(leng)  && aName.before(leng) == dir) {
		    // The leading ./ indicates that directory is removed.
		    return "./" + aName.from (leng);
		}
	    }
	}
    }
    if (leng == 0) {
	// Also relative if the parent is this directory and the
	// subtable is relative.
	if (name[0] != '/'  &&  name[0] != '$'  &&  name[0] != '~') {
	    return "./" + name;
	}
    }
    // Nothing could be stripped.
    // Return original name after removing possible leading ./
    String tName(name);
    while (tName.length() >= 2  &&  tName[0] == '.'  &&  tName[1] == '/') {
	tName = tName.after(1);
    }
    return tName;
}

String Path::addDirectory (const String& name, const String& otherName)
{
    // If a directory was stripped off, add the directory of the parent.
    // Otherwise use the name as such.
    // Note that dirName will use the expanded name.
    // Start with removing possible leading ./
    String tName(name);
    while (tName.length() >= 2  &&  tName[0] == '.'  &&  tName[1] == '/') {
	tName = tName.after(1);
    }
    // If anything was removed, we have to add the directory.
    if (tName.length() < name.length()) {
	// If ./ was removed, add directory of otherName.
	// Otherwise add the full otherName.
	Path dir(otherName);
	if (tName.length() == name.length() - 2) {
	    dir = Path(dir.dirName());
	}
	dir.append (tName);
	return dir.originalName();
    } else {
        // If name ends in /. and if it matches the end of otherName,
        // we have to return the remainder of otherName.
        Int leng = tName.length();
        if (leng >= 3  &&  tName[leng-2] == '/'  &&  tName[leng-1] == '.') {
	    leng -= 2;
	    String oName(otherName);
	    Int oleng = oName.length();
	    if (oleng >= leng+2
            &&  '/'+tName.before(leng) == oName.from(oleng-leng-1)) {
	        return oName.before(oleng-leng-1);
	    }
	}
    }
    return tName;
}

} //# NAMESPACE CASA - END

