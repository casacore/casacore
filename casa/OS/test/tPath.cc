//# tPath.cc: Test program for class Path
//# Copyright (C) 1993,1994,1995,1996,1998,1999,2000,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <casa/OS/Path.h>
#include <casa/OS/EnvVar.h>
#include <casa/BasicSL/String.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions.h>
#include <casa/iostream.h>
#include <unistd.h>


#include <casa/namespace.h>
// <summary>
// Test program for class Path.
// </summary>

// This program tests the class Path.
// The results are written to stdout. The script executing this program,
// compares the results with the reference output file.
// <p>
// When an argument is given, no exceptions will be thrown.
// This can be used to check if no memory leaks occur in normal operation.


void check (const String& path, const String& expanded,
	    const String& absolute, Bool& success)
{
    cout << ">>> testing path = " << path << endl;
    cout << "<<<" << endl;
    Path test (path);
    const String& ov = test.originalName();
    if (ov != path) {
	cout << "original: " << ov << endl;
	cout << "expected: " << path << endl;
	success = False;
    }
    const String& ev = test.expandedName();
    if (ev != expanded) {
	cout << "expanded: " << ev << endl;
	cout << "expected: " << expanded << endl;
	success = False;
    }
    const String& av = test.absoluteName();
    if (av != absolute) {
	cout << "absolute: " << av << endl;
	cout << "expected: " << absolute << endl;
	success = False;
    }
}

void checkDirBase (const String& path, const String& dir,
		   const String& base, Bool& success)
{
    cout << "testing dirbase path = " << path << endl;
    Path test (path);
    String dv = test.dirName();
    if (dv != dir) {
	cout << "dir:      " << dv << endl;
	cout << "expected: " << dir << endl;
	success = False;
    }
    String bv = test.baseName();
    if (bv != base) {
	cout << "base:     " << bv << endl;
	cout << "expected: " << base << endl;
	success = False;
    }
}

void doIt (Bool doExcp, Bool& success)
{
    // Get the home directory.
    String home (EnvironmentVariable::get ("HOME"));
    AlwaysAssertExit (! home.empty());
    // Get the current working directory (set in tPath.exec).
    String curr (EnvironmentVariable::get ("tPath_Env_Curr"));
    AlwaysAssertExit (! curr.empty());
    // Get the user name.
    String user (EnvironmentVariable::get ("tPath_Env_Username"));
    AlwaysAssertExit (! user.empty());

    // Define all kind of path names (relative, absolute, with and
    // without environment variables and tilde).
    // Test the resulting expanded and absolute name.
    check ("~", home, home, success);
    check ("$HOME", home, home, success);
    check ("/$HOME/", "/"+home+"/", home, success);
    check ("~/testx", home+"/testx", home+"/testx", success);
    check ("$HOME/testx", home+"/testx", home+"/testx", success);
    check ("~" + user, home, home, success);
    check ("~" + user + "/test/test2", home + "/test/test2",
	   home + "/test/test2", success);
    check ("$tPath_Env_User/test", home + "/test", home + "/test", success);
    check ("${tPath_Env_User}/test", home + "/test", home + "/test", success);
    check ("/aa$HOME/test", "/aa" + home + "/test",
           "/aa" + home + "/test", success);
    check ("/aa${HOME}bb/test", "/aa" + home + "bb/test",
           "/aa" + home + "bb/test", success);
    check ("/aa${HOME}bb${HOME}cc/test", "/aa"+ home+"bb"+home+"cc/test",
           "/aa"+home+"bb"+home+"cc/test", success);
    check ("/testx", "/testx", "/testx", success);
    check ("testx", "testx", curr + "/testx", success);
    check (user, user, curr + "/" + user, success);
    check ("$tPath_Env_Test1/$HOME", home + "/" +  home,
	   home + home, success);

    // Check resolvedName.
    String tpDir = Path("$tPath_Env_Curr/tPath_tmpdir").absoluteName();
    AlwaysAssertExit (Path("$tPath_Env_Curr/tPath_tmpdir//d1/../d1/.//d2/").
                      resolvedName() == tpDir + "/d1/d2");
    if (doExcp) {
      Bool ok = True;
	try {
            Path("/a/b").resolvedName();
	} catch (AipsError x) {
            cout << ">>> " << x.getMesg() << endl << "<<<" << endl;
            ok = False;
	}
        AlwaysAssertExit (!ok);
    }

    // Check copy ctor and assignment (also self-assignment).
    String str;
    Path test1 ("~");
    Path test2 (test1);
    AlwaysAssertExit (test1.originalName() == test2.originalName());
    AlwaysAssertExit (test1.expandedName() == test2.expandedName());
    AlwaysAssertExit (test1.absoluteName() == test2.absoluteName());
    Path test3 ("~/test");
    str = test3.expandedName();
    str = test3.absoluteName();
    test3 = test2;
    AlwaysAssertExit (test3.originalName() == test2.originalName());
    AlwaysAssertExit (test3.expandedName() == test2.expandedName());
    AlwaysAssertExit (test3.absoluteName() == test2.absoluteName());
    test3 = test3;
    AlwaysAssertExit (test3.originalName() == test2.originalName());
    AlwaysAssertExit (test3.expandedName() == test2.expandedName());
    AlwaysAssertExit (test3.absoluteName() == test2.absoluteName());
    
    // Testing exception because of recursive environment variables.
    EnvironmentVariable::set ("TEST1", "$TEST2");
    EnvironmentVariable::set ("TEST2", "$TEST1");
    test1 = Path ("$TEST1");
    if (doExcp) {
	try {
	    test1.expandedName ();
	} catch (AipsError x) {
	    cout << x.getMesg () << endl;
	} 
    }
	
    // Tests for isValid and isStrictlyPosix
    test1 = Path ("HOME//test");
    test2 = Path ("");
    test3 = Path ("\033");
    AlwaysAssertExit (! test1.isValid());
    AlwaysAssertExit (test2.isValid());
    AlwaysAssertExit (! test3.isValid());
    AlwaysAssertExit (! test1.isStrictlyPosix());
    AlwaysAssertExit (test2.isStrictlyPosix());
    AlwaysAssertExit (! test3.isStrictlyPosix());
    test1 = Path ("$HOME/test");
    AlwaysAssertExit (test1.isValid());
    AlwaysAssertExit (test1.isStrictlyPosix());

    cout << test1.length() << endl;
    cout << ">>> ";
    cout << test1.maxLength() << endl;
    cout << "<<<" << endl;

    // Tests for dir- and baseName.
    checkDirBase (".", ".", ".", success);
    checkDirBase ("..", ".", "..", success);
    checkDirBase ("a", ".", "a", success);
    checkDirBase ("a/", ".", "a", success);
    checkDirBase ("/b", "/", "b", success);
    checkDirBase ("/b/", "/", "b", success);
    checkDirBase ("/", "/", "", success);
    checkDirBase ("a/b", "a", "b", success);
    checkDirBase ("a/b/", "a", "b", success);
    checkDirBase ("~/test/test1", home + "/test", "test1", success);
    checkDirBase ("/home/", "/", "home", success);
    checkDirBase ("test/test1/..", "test/test1", "..", success);
    checkDirBase ("harry", ".", "harry", success);
    checkDirBase ("/home", "/", "home", success);
    checkDirBase ("$HOME/test/ha./test", home + "/test/ha.", "test", success);

    Path tpath;
    AlwaysAssertExit (tpath.originalName() == ".");
    tpath = Path("");
    AlwaysAssertExit (tpath.originalName() == ".");
    tpath = Path("/");
    AlwaysAssertExit (tpath.originalName() == "/");
    tpath.append ("abc/");
    AlwaysAssertExit (tpath.originalName() == "/abc/");
    tpath.append ("/abc");
    AlwaysAssertExit (tpath.originalName() == "/abc//abc");
    AlwaysAssertExit (tpath.absoluteName() == "/abc/abc");
    tpath.append ("abc");
    AlwaysAssertExit (tpath.originalName() == "/abc//abc/abc");
    tpath.append ("/abc");
    AlwaysAssertExit (tpath.originalName() == "/abc//abc/abc/abc");
    AlwaysAssertExit (tpath.absoluteName() == "/abc/abc/abc/abc");

    // Test the strip/addDirectory functionality.
    AlwaysAssertExit (Path::stripDirectory ("././abc", "././././abc/de")
		      == "de/.");
    AlwaysAssertExit (Path::stripDirectory ("././abc", "././././abd/de")
		      == "abc");
    AlwaysAssertExit (Path::stripDirectory ("abc", "abc/de")
		      == "de/.");
    AlwaysAssertExit (Path::stripDirectory ("././abc/de/ef", "././././abc/de")
		      == "././ef");
    AlwaysAssertExit (Path::stripDirectory ("abc/ef", "abc/de")
		      == "./ef");
    AlwaysAssertExit (Path::stripDirectory ("ef", "de")
		      == "./ef");
    AlwaysAssertExit (Path::stripDirectory ("$HOME/ef", "de")
		      == "$HOME/ef");
    AlwaysAssertExit (Path::stripDirectory ("/de/abc/ab/ef", "/de/abc/de")
		      == "./ab/ef");
    AlwaysAssertExit (Path::stripDirectory ("/de/abc/ab/ef", "/de/abc/ab")
		      == "././ef");
    AlwaysAssertExit (Path::addDirectory ("de/ab", "/ghi/a")
		      == "de/ab");
    AlwaysAssertExit (Path::addDirectory ("./de/ab", "/ghi/a")
		      == "/ghi/de/ab");
    AlwaysAssertExit (Path::addDirectory ("././de/ab", "/ghi/a")
		      == "/ghi/a/de/ab");
}


int main (int argc, const char*[])
{
    Bool success = True;
    try {
	doIt ( (argc<2), success);
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } 
    if (success) {
	cout << "OK" << endl;
	return 0;                           // exit with success status
    }
}
