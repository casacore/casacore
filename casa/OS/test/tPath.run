#! /bin/sh
#-----------------------------------------------------------------------------
# Script to execute the tPath program via assay.
# It defines a few environment variables which are used in tPath.
# Execution:
#      tPath.run        execute using assay (normal procedure)
#      tPath.run 0      only define the subdirectory, etc. (for debugging)
#      tPath.run 1      execute as such (to test)
#      tPath.run 1 1    execute without exceptions (to test leakage)
#=============================================================================

tPath_Env_Username=`whoami`
export tPath_Env_Username
tPath_Env_User="~`whoami`"
export tPath_Env_User
tPath_Env_Curr=`pwd`
export tPath_Env_Curr
tPath_Env_Test1='$HOME'
export tPath_Env_Test1

if [ $# != 0 ]
then
    if [ $1 != 0 ]
    then
        ./tPath $2
    fi
else
    ./tPath
fi
