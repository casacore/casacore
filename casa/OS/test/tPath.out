>>> testing path = ~
<<<
>>> testing path = $HOME
<<<
>>> testing path = /$HOME/
<<<
>>> testing path = ~/testx
<<<
>>> testing path = $HOME/testx
<<<
>>> testing path = ~gvandiep
<<<
>>> testing path = ~gvandiep/test/test2
<<<
>>> testing path = $tPath_Env_User/test
<<<
>>> testing path = /testx
<<<
>>> testing path = testx
<<<
>>> testing path = gvandiep
<<<
>>> testing path = $tPath_Env_Test1/$HOME
<<<
Path::expandName: recursive environment variable
10
>>> 1024
<<<
testing dirbase path = .
testing dirbase path = ..
testing dirbase path = a
testing dirbase path = a/
testing dirbase path = /b
testing dirbase path = /b/
testing dirbase path = /
testing dirbase path = a/b
testing dirbase path = a/b/
testing dirbase path = ~/test/test1
testing dirbase path = /home/
testing dirbase path = test/test1/..
testing dirbase path = harry
testing dirbase path = /home
testing dirbase path = $HOME/test/ha./test
OK
